### DISCUSSION TITLE
Querying REDCap database fails in targets pipeline

### DISCUSSION THREAD

**Discussion by Scott-Coggeshall** [ropensci/targets#625] [Help] [answered]
I have data in a REDCap database that I want to pull into R via the REDCap API using the `redcap_read` function from the `REDCapR` package. If I run something like the following in a normal R session it works fine:

```
my_data <- REDCapR::redcap_read(redcap_uri = "https://fakeredcapdb.com/api/", token = "mytoken")
```

(with values for `redcap_uri` and `token` replaced by the actual uri and API token). 

I want to do this within a targets plan like the following

```
library(targets)

tar_option_set(packages = c("REDCapR"))

list(
  tar_target(my_uri, "https://fakeredcapdb.com/api/", format = "url"),
  tar_target(my_token, "mytoken"),
  tar_target(data, redcap_read(redcap_uri = my_uri, token = my_token)))
)

```

However, running `tar_make()` results in the following error:

```
Error : HTTP response status code 501
Could not access url:
  https://fakeredcapdb.com/api/
HTTP response headers:
  cache-control = no-store, no-cache, must-revalidate
  pragma = no-cache
  content-length = 0
  content-type = text/xml; charset=utf-8
  expires = 0
  server = Microsoft-IIS/8.5
  x-xss-protection = 1; mode=block
  x-content-type-options = nosniff
  access-control-allow-origin = *
  x-powered-by = ASP.NET
  date = Thu, 02 Sep 2021 18:47:38 GMT

```

Do you have any idea why I'm able to access the url outside of a targets plan but not inside of one?

**Comment by wlandau** [comment, marked as answer]
`format = "url"` is for when you have a specific file in mind at a URL, e.g. https://raw.githubusercontent.com/wlandau/targets-minimal/main/data/raw_data.csv. For APIs, I would recommend something like:

```r
tar_target(
  name = data,
  command = redcap_read(token = Sys.getenv("YOUR_TOKEN")),
  format = "feather" # assuming data frame format
  cue = tarchetypes::tar_cue_force(redcap_changed(Sys.getenv("YOUR_TOKEN")))
)
```

where `redcap_changed()` is a function you write to check if the remote database changed (using the last modified time stamp or hash of the database table). [`tarchetypes`](https://docs.ropensci.org/tarchetypes/reference/index.html) has a bunch of alternatives for this sort of thing, such as `tar_change()`, `tar_age()`, and `tar_cue_age()`.

**Comment by Scott-Coggeshall** [comment]
Sorry for the late reply. This solved the issue, thanks!
